[frameit] Support for Android

milhauscz opened the pull request:
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context

Resolves #14968

🔑This PR adds Android support to frameit. [Based on this discussion.](https://github.com/fastlane/fastlane/issues/14968) [It is related to this PR which adds new Android frames.](https://github.com/fastlane/frameit-frames/pull/10)

Originally, frameit only supported iOS devices and relied heavily on device resolutions to detect individual devices and using "Space Gray" as the default color, later adding more and more exceptions and if / else branches to solve conflicts. It was also depending on the "deliver" module which is exclusive for iOS (Android uses the "supply" module for uploading screenshots). This PR contains a lot of refactoring, mainly:
- Frameit is decoupled from the deliver module now
- New Device class which contains all necessary device information: id, device name, priority, resolutions, screen pixel density, default color and platform
- New algorithm for detecting device: each device has an assigned priority. Device is selected according to resolution, priority and platform. Priority can be modified by adding device name to the screenshot name, using force_device_type option or the old use_legacy_iphone options. Platform can be specified by the new "use_platform" parameter (see below).
- If a device with a color specified by the user cannot be found, the frame with the device's default color is used.
- Added "use_platform" parameter which can be specified in Framefile.json, fastfile or passed via CLI parameters; default is iOS for backward compatibility
- Added new "android" and "ios" commands for frameit
- Updated frames_generator to include Android devices and exclude some deprecated devices (separate PR has been created)
- Updated frames_generator's image magick commands to remove shadows when checking the device offsets to support Android devices (specifically Samsung Galaxy S9 whose screen contains shadows representing round edges could not be processed without this change). This change is backward compatible with Apple devices.
- Updated frameit docs with Android support
- Created unit tests for the new Device class
- Fixed force_device_type option which didn't work before, also added support to Framefile.json
- Added iPhone 8 and iPad Mini 4 support
- Added new option :use_legacy_iphone7 to prefer iPhone 7 over iPhone 8
- Code style fixes according to rubocop

How the PR was tested:
- Created new unit tests for the Device class
- Successfully ran all the unit tests
- Successfully tested several scenarios using the Framefile.json and use_platform, frame, filter and force_device_type options, tested backward compatibility with the use_legacy_iphone options. Tested the new ios and android commands.

milhauscz committed:
Adding Android support to frameit
Frameit refactoring and decoupling from the deliver module
New Device class which contains all necessary device information: id, device name, priority, resolutions, screen pixel density, default color and platform
Added "use_platform" parameter which can be specified in Framefile.json, fastfile or passed via CLI parameters; default is iOS for backward compatibility
Added new "android" and "ios" commands for frameit
Updated frames_generator to include Android devices and exclude some deprecated devices
Updated frames_generator's image magick commands to remove shadows and correctly process Android devices (specifically Samsung Galaxy S9 whose screen contains shadows representing round edges)
Updated frameit docs with Android support
Created unit tests for the new Device class
Fixed force_device_type option which didn't work before, also added support to Framefile.json
Added iPhone 8 support
Added new option :use_legacy_iphone7 to prefer iPhone 7 over iPhone 8
Code style fixes according to rubocop

janpio commented on frameit/frames_generator/Rakefile:
```diff
@@ -46,9 +72,10 @@ task(:generate_device_frames) do
   image_assets = []
   devices = Hash.new { |h, k| h[k] = [] }
-  (Dir["raw/Phones/Apple*/Device/*.png"] + Dir["raw/Computers/Apple*/Device/*.png"] + Dir["raw/Tablets/Apple*/Device/*.png"]).each do |current_raw|
+  # According to mac_editor.rb, Mac screenshots do not need frames - do we need to process the Computers folder?
```
Yeah, having the option to frame computers still seems like a good thing to have (or at least having the frames, to one day write the actual code for that if it is currently not (properly) implemented)

milhauscz replied:
Understood, so the comment could be changed to this:
```suggestion
  # Frames from the "computers" folder are currently not used by the mac_editor.rb, but may be used in the future if implemented.
```

janpio replied:
Either that, or just remove completely.

milhauscz replied:
OK, removed completely in the new commit.

janpio commented on frameit/frames_generator/Rakefile:
```diff
@@ -98,13 +125,20 @@ task(:generate_device_frames) do
 
 def sanitize_filename(filename)
-  filename.gsub('Grey', 'Gray')
+  filename.gsub('Grey', 'Gray').gsub(' - ', ' ')
```
What is the ` - ` replacement used for?

milhauscz replied:
This is because the naming of the devices from Facebook is inconsistent and the Google Pixel devices contain a dash between their name and color, e. g. "Google Pixel 3 XL - Clearly White.png". We don't want to have that dash in the device name.

janpio replied:
Ok, best add that as a comment (probably same for the Grey/Gray thing).

milhauscz replied:
OK, added a comment in the new commit.

janpio commented on frameit/frames_generator/Rakefile:
```diff
@@ -124,6 +158,16 @@ def measure_slot(path)
     command_output = MiniMagick::Tool::Magick.new do |magick|
       magick << path
+
+      # some devices (e. g. Samsung Galaxy S9) may have shadows inside their screen to show the round edges - we need
```
This (and following lines) looks like a possible reason for the changed height of iOS frames from https://github.com/fastlane/frameit-frames/pull/10#issuecomment-514535858. Could you test locally if removing this logic here adds that 1px back? If so, this here could possibly have negative effect on iOS (will have to look into the exact details on the iOS differences to identify if this is really what is happening.)

milhauscz replied:
These lines didn't change the height, because these are in the measure_slot function which only modifies the image in memory when searching the offsets which are subsequently saved to offsets.json. The code which modifies and saves the frames is in convert_image_resource. I still tested it according to your instructions and the generated size is still off by 1 PX. 

However, I found something suspicious. [In this pull request](https://github.com/fastlane/fastlane/pull/14199), a change in frames_generator has been introduced which should avoid these 1 - 2 PX differences. This PR was merged on February 20, but last commit in the frameit-frames repo was on January 19. Is it possible the frames were not generated with the new code yet and my commit is the first one which brings these fixed sizes to the frameit-frames repo?

janpio replied:
Yes, that makes total sense. Let's pull @mollyIV in here to confirm.

milhauscz replied:
I've just tried removing the "co.fuzz("90%")" line and generating frames. Apple iPad Air 2 Gold had 2496 px height instead of 2495 px. So that's it.

janpio commented:
Let me start by expressing how awesome the PR description makes this sound ❤️ And now on to understand (and possibly nitpick) the code ;)

janpio commented on fastlane/lib/fastlane/actions/docs/frame_screenshots.md:
```diff
@@ -129,8 +136,10 @@ The general parameters are defined in the `default` key and can be:
 | `interline_spacing` | Specifies whether _frameit_ should add or subtract this many pixels between the individual lines of text. This only applies to a multi-line `title` and/or `keyword` to expand or squash together the individual lines of text. | `0` |
 | `font_scale_factor` | Specifies whether _frameit_ should increase or decrease the font size of the text. | `0.1` |
-| `frame` | Overrides the color of the frame to be used. (Valid values are `BLACK, `WHITE`, `GOLD` and `ROSE_GOLD`) | NA |
+| `frame` | Overrides the color of the frame to be used. (Valid values are `BLACK`, `WHITE`, `GOLD` and `ROSE_GOLD`) | NA |
```
Is this still limited to these options or can the new colors be used here as well?

milhauscz replied:
Yes, I forgot to update this part of the documentation. Shall I add the list of all supported colors, maybe separating iOS and Android colors?

janpio replied:
That is actually a lot more complicated now, because when this was first written all colors were available for all devices. This obviously hasn't been true for quite some time now.

But that shouldn't be your problem, so we can decide now if it's better to just add them to the list or leave them off - but create a new issue so this gets fixed and documented better (simplest solution I can think of right now is for this just linking to a "ABout device colors" headline that then has a separated list for iOS and Android, or maybe even a table which devices are available in what color.

milhauscz replied:
OK, I have created a feature request. #15089

janpio commented on fastlane/lib/fastlane/actions/docs/frame_screenshots.md:
```diff
@@ -129,8 +136,10 @@ The general parameters are defined in the `default` key and can be:
+| `frame` | Overrides the color of the frame to be used. (Valid values are `BLACK`, `WHITE`, `GOLD` and `ROSE_GOLD`) | NA |
 | `title_min_height` | Specifies a height always reserved for the title. Value can be a percentage of the height or an absolute value. The device will be placed below (or above) this area. Convenient to ensure the device top (or bottom) will be consistently placed at the same height on the different screenshots. | NA |
+| `use_platform` | Overrides the platform used for the screenshot. Valid values are `IOS`, `ANDROID` and `ANY`. | `IOS` |
+| `force_device_type` | Forces a specific device. Valid values are: Huawei P8, Motorola Moto E, Motorola Moto G, Nexus 4, Nexus 5X, Nexus 6P, Nexus 9, Samsung Galaxy Grand Prime, Samsung Galaxy Note 5, Samsung Galaxy S Duos, Samsung Galaxy S3, Samsung Galaxy S5, Samsung Galaxy S7, Samsung Galaxy S8, Samsung Galaxy S9, iPhone 5s, iPhone 5c, iPhone SE, iPhone 6s, iPhone 6s Plus, iPhone 7, iPhone 7  ...
```
Can we generate this list of devices somehow? Might have to change the simple .md to a file that can execute code though, I think we have that somewhere else already, .md.erb or similar.

milhauscz replied:
I copied this by running frameit with a wrong parameter in force_device_type. The generated error message contains all these names. Of course some automatization would be better in the future so that this doesn't have to be updated every time a new device is added. I will check how this can be done.

janpio replied:
I was thinking of https://github.com/fastlane/fastlane/blob/294a80d946c24b82096b05c1fb0c0d91157ac3b0/fastlane/lib/fastlane/actions/docs/upload_to_app_store.md.erb#L455-L497 - but feel free to declare this out of scope here and handle this in a follow up PR.

milhauscz replied:
All right, thanks for the information, I put this to my notes and will get to it later.

janpio commented on frameit/lib/frameit/device_types.rb:
```diff
@@ -1,18 +1,98 @@
+    SUNRISE_GOLD = "Sunrise Gold"
+    TITANIUM_GRAY = "Titanium Gray"
+    CORAL = "Coral"
+    BLACK = "Black"
```
Ok, before Black was Space Gray.. we have to be aware of that.

milhauscz replied:
Yes I know. Before BLACK was Space Gray (or Matte Black in one device) and it was the default color. Now the default color is nil and each device has its own default color, for all Apple devices I assigned Space Gray or Matte Black if available. So it should be backward compatible, just if someone explicitly required "BLACK", it will show him a message like "Black not available, falling back to Space Gray".

milhauscz replied:
By the "nil" I mean if template_finder receives "nil" as the color, it will use the particular device's default color, i. e. space gray or matte black in Apple devices.

janpio replied:
Ok, defining the default color per device become possible as you created this new list of devices and their properties here for frameit, correct?

Automatic fallback is perfect and makes this a great and acceptable change.

milhauscz replied:
Yes, each device is assigned a color constant from this "enum" which serves as a fallback. This avoids a lot of if / else code. Thanks.

janpio commented on frameit/lib/frameit/device_types.rb:
```diff
@@ -1,18 +1,98 @@
+    end
+  end
+
+  module Devices
```
The iOS devices information here is "duplicated" from `deliver`, correct?

milhauscz replied:
Yes, except iPad, which is separated into four constants in Deliver containing some iPad versions whose frames we don't have and some of them having some weird / unclear resolutions. So in case of iPad I checked the GSM Arena information and double checked by manually measuring the frame's inner size.

milhauscz replied:
One more thing - there is a new kind of information: pixel density. This information is not in `deliver` and was copied from the GSM Arena website. It is used in `Screenshot` for determining whether the device is mini or triple density. Before there were just if clauses listing devices, now the exact pixel density is checked and gives the same result as before for Apple devices.

janpio commented:
Ok, first round is done. Will have to spend loads of more time understanding this - especially trying to understand if this could somehow change existing behavior.

milhauscz committed:
Modified some frames_generator comments according to the discussion

milhauscz commented:
> Ok, first round is done. Will have to spend loads of more time understanding this - especially trying to understand if this could somehow change existing behavior.

Sure, the changes are extensive, let me know if something is unclear.

milhauscz commented:
I've just noticed #14787 which adds support for iPad Pro 12.9 (3rd Generation). I saw this in code, but since the frameit-frames repo didn't contain frame for this device, I didn't include it in the [Devices enum in device_types.rb](https://github.com/milhauscz/fastlane/blob/master/frameit/lib/frameit/device_types.rb). I thought it is just a device supported by the deliver module but for which we don't have a frame, therefore shouldn't be included in frameit. This may apply to some other iPad versions. I didn't know it is frameit users' common practice to [add custom frames into their frameit frames folder](https://github.com/fastlane/fastlane/pull/14787#issuecomment-494749770). In that case I can just add these missing devices to the Devices enum. To sum it up, regarding iPad the frameit-frames currently contains:

- Apple iPad Air 2 (not referenced anywhere in deliver/app_screenshots.rb, therefore by adding it to the Devices enum I also added support for it)
- Apple iPad Mini 4 (also not referenced in deliver)
- Apple iPad Pro (the only one referenced in deliver)

[Deliver specifies these screen sizes ](https://github.com/fastlane/fastlane/blob/master/deliver/lib/deliver/app_screenshot.rb)and their respective device names, which were used for retrieving frames:

- ScreenSize::IOS_IPAD => "iPad"
- ScreenSize::IOS_IPAD_10_5 => "iPad 10.5",
- ScreenSize::IOS_IPAD_11 => "iPad 11",
- ScreenSize::IOS_IPAD_PRO => "iPad Pro",
- ScreenSize::IOS_IPAD_PRO_12_9 => "iPad Pro (12.9-inch) (3rd generation)"

So I would have to add iPad, iPad 10.5, iPad 11 and iPad Pro (12.9-inch) (3rd generation) to Devices with lower priorities than the ones whose frames are included in frameit-frames repo. What do you think?

milhauscz commented:
I've just noticed #14787 which adds support for iPad Pro 12.9 (3rd Generation). I saw this in code, but since the frameit-frames repo didn't contain frame for this device, I didn't include it in the Devices enum in device_types.rb. I thought it is just a device supported by the deliver module but for which we don't have a frame, therefore shouldn't be included in frameit. I didn't know it is frameit users' common practice to add custom frames into their frameit frames folder. In that case I can just add this missing device to the Devices enum. To sum it up, regarding iPad the frameit-frames currently contains:

- Apple iPad Air 2
- Apple iPad Mini 4 (not referenced anywhere in [frameit's screenshot.rb] / device_name (https://github.com/fastlane/fastlane/blob/master/frameit/lib/frameit/screenshot.rb), therefore by adding it to the Devices enum I also added support for it)
- Apple iPad Pro

The new device "iPad Pro (12.9-inch) (3rd generation)" with "ipadPro129" ID will be added to the enum so that the PR will be backward compatible with #14787. Also will modify the algorithm so that it will search for both ID and device name in the file name.

milhauscz committed:
Added iPad Pro (12.9-inch) (3rd generation) to ensure backward compatibility with PR #14787

milhauscz committed:
Added missing "Apple" to the iPad 12.9 inch 3rd generation name
Searching algorithm compares filename with the device's name without "Apple" for backward compatibility

janpio commented:
Good observation, and I think correct conclusion and solution.
We will need to work on keeping both locations up to data. Maybe a test that makes sure that all the deliver devices are also supported in frameit?

milhauscz commented:
> Good observation, and I think correct conclusion and solution.
> We will need to work on keeping both locations up to data. Maybe a test that makes sure that all the deliver devices are also supported in frameit?

Actually not all devices from `deliver` are supported in `frameit`. The devices supported by `deliver` can be [found in app_screenshot.rb](https://github.com/fastlane/fastlane/blob/master/deliver/lib/deliver/app_screenshot.rb), devices supported by `frameit` in the [original screenshot.rb / device_name](https://github.com/fastlane/fastlane/blob/master/frameit/lib/frameit/screenshot.rb) function which maps `deliver`'s enum to device names which are used for retrieving the actual frames (just 'Apple' is prepended and '[color].png' appended). For example, I can see there is IOS_IPAD_10_5 in `deliver`, but it is not in `frameit`. But after adding the missing iPad it seems we got them all covered.

mollyIV commented on frameit/lib/frameit/device_types.rb:
```diff
@@ -1,18 +1,102 @@
+    # ID and formatted name must be exactly as specified so that device.detect_device() will select this device if the filename includes them
+    IPAD_PRO_12_9 = Frameit::Device.new("ipadPro129", "Apple iPad Pro (12.9-inch) (3rd generation)", 2, [[2048, 2732], [2732, 2048]], 264, Color::SPACE_GRAY, Platform::IOS, Deliver::AppScreenshot::ScreenSize::IOS_IPAD_PRO_12_9)
+    MAC = Frameit::Device.new("mac", "Apple MacBook", 0, [[1280, 800], [1440, 900], [2560, 1600], [2880, 1800]], nil, Color::SPACE_GRAY, Platform::IOS, Deliver::AppScreenshot::ScreenSize::MAC)
+  end
```
Hmm, wondering if it's worth to add unit tests to check a given device configuration (even though it'd be mostly checking setters). WDYT? 🤔

milhauscz replied:
If an ID or name changes in the future, it could cause a breaking change. On the other hand if a developer unfamiliar with `frameit` deliberately changes a value, nothing stops him from modifying the respective unit test to pass it. So I think relying on thorough PR reviews would be better.

mollyIV commented on frameit/lib/frameit/editor.rb:
```diff
@@ -488,16 +489,9 @@ def build_text_images(max_width, max_height, stack_title)
 
-    # Loads the config (colors, background, texts, etc.)
     # Don't use this method to access the actual text and use `fetch_texts` instead
     def fetch_config
```
Are we still using this method? It simply returns `@config` property now, so maybe we could just expose a public getter of it instead of having an extra method?

milhauscz replied:
You're right, I will refactor all `fetch_config` to `@config`.

milhauscz replied:
[Removed the function in the new commit.](https://github.com/milhauscz/fastlane/commit/91927056a98b8def508b2da405020c105676dfab)

mollyIV commented on frameit/lib/frameit/template_finder.rb:
```diff
@@ -11,32 +10,36 @@ class TemplateFinder
-
-      templates = Dir["#{FrameDownloader.templates_path}/#{filename}.{png,jpg}"] # ~/.frameit folder
+      filename = create_file_name(screenshot.device_name, screenshot.color.nil? ? screenshot.default_color : screenshot.color)
+      templates = Dir["#{FrameDownloader.templates_path}/#{filename}.{png,jpg}"] # ~/.frameit_android folder
```
`~/.frameit_android` why is this one Android specific?

milhauscz replied:
I forgot to revert this change. In the beginning I was making the frameit with Android as a separate module. Will revert in a new commit.

milhauscz replied:
[Fixed in the new commit.](https://github.com/milhauscz/fastlane/commit/e52dafa0a43f9306e0ad8245e81ceac5a0550065)

mollyIV reviewed (changes requested):
Hey @milhauscz 👋 

First of all thank you so much for your work, we really appreciate it 🙇 Supporting Android seems to be a great addition to `fastlane` and I believe many will benefit from it 🎊 

Added a few comments, please let me know if something's not clear. 

I would like to test your changes. You did mention that:

> Successfully tested several scenarios using the Framefile.json and use_platform, frame, filter and force_device_type options, tested backward compatibility with the use_legacy_iphone options. Tested the new ios and android commands.

do you have a demo project that we could play with? 🙏 

In the future it would be really great to have iOS and Android parts of the `frameit` separated and have a `core` part being reused between those, wouldn't be? 😉

janpio commented on fastlane/lib/fastlane/actions/docs/frame_screenshots.md:
```diff
@@ -23,9 +23,10 @@ _frameit_ allows you to put a gorgeous device frame around your iOS and macOS sc
+Put a gorgeous device frame around your iOS, macOS and Android screenshots just by running one simple command. Support for:
 
 - iPhone, iPad and Mac
+- Select Android devices
```
Or
```suggestion
- Set of Android devices
```

janpio commented on fastlane/lib/fastlane/actions/docs/frame_screenshots.md:
```diff
@@ -129,8 +136,10 @@ The general parameters are defined in the `default` key and can be:
+| `frame` | Overrides the color of the frame to be used. (Valid values are `BLACK`, `WHITE`, `GOLD` and `ROSE_GOLD`) | NA |
 | `title_min_height` | Specifies a height always reserved for the title. Value can be a percentage of the height or an absolute value. The device will be placed below (or above) this area. Convenient to ensure the device top (or bottom) will be consistently placed at the same height on the different screenshots. | NA |
+| `use_platform` | Overrides the platform used for the screenshot. Valid values are `IOS`, `ANDROID` and `ANY`. | `IOS` |
+| `force_device_type` | Forces a specific device. Valid values are: Huawei P8, Motorola Moto E, Motorola Moto G, Nexus 4, Nexus 5X, Nexus 6P, Nexus 9, Samsung Galaxy Grand Prime, Samsung Galaxy Note 5, Samsung Galaxy S Duos, Samsung Galaxy S3, Samsung Galaxy S5, Samsung Galaxy S7, Samsung Galaxy S8, Samsung Galaxy S9, iPhone 5s, iPhone 5c, iPhone SE, iPhone 6s, iPhone 6s Plus, iPhone 7, iPhone 7  ...
```
The ones supplied by Facebook Design

milhauscz commented on frameit/lib/frameit/device.rb:
```diff
@@ -0,0 +1,102 @@
+      # prefer filename
+      unless filename_device.nil?
+        return filename_device
+      end
```
I agree, I made this code because I am used to Java where I usually do something like this:

`if (filenameDevice != null) return filenameDevice;`

And later changed it to the code above because of rubocop requiring using `.nil?` instead.

milhauscz replied:
[Fixed in the new commit.](https://github.com/milhauscz/fastlane/commit/d92a21b6c2cdce99009cd7f1b826ccc20c3a5fae)

milhauscz commented on fastlane/lib/fastlane/actions/docs/frame_screenshots.md:
```diff
@@ -129,8 +136,10 @@ The general parameters are defined in the `default` key and can be:
-| `frame` | Overrides the color of the frame to be used. (Valid values are `BLACK, `WHITE`, `GOLD` and `ROSE_GOLD`) | NA |
+| `frame` | Overrides the color of the frame to be used. (Valid values are `BLACK`, `WHITE`, `GOLD` and `ROSE_GOLD`) | NA |
 | `title_min_height` | Specifies a height always reserved for the title. Value can be a percentage of the height or an absolute value. The device will be placed below (or above) this area. Convenient to ensure the device top (or bottom) will be consistently placed at the same height on the different screenshots. | NA |
+| `use_platform` | Overrides the platform used for the screenshot. Valid values are `IOS`, `ANDROID` and `ANY`. | `IOS` |
```
Correct, `ANY` means frameit will look for the device frame in the whole "enum", i. e. the platform filtering will be switched off.

milhauscz commented on frameit/lib/frameit/device.rb:
```diff
@@ -0,0 +1,102 @@
+
+    def priority
+      if !priority_config_key.nil? && Frameit.config[priority_config_key]
+        999
```
OK, let's create a constant and we can also create a check in the Device constructor to disallow priorities greater than this constant. Will modify in a new commit.

milhauscz commented:
> Hey @milhauscz 👋
> 
> First of all thank you so much for your work, we really appreciate it 🙇 Supporting Android seems to be a great addition to `fastlane` and I believe many will benefit from it 🎊
> 
> Added a few comments, please let me know if something's not clear.
> 
> I would like to test your changes. You did mention that:
> 
> > Successfully tested several scenarios using the Framefile.json and use_platform, frame, filter and force_device_type options, tested backward compatibility with the use_legacy_iphone options. Tested the new ios and android commands.
> 
> do you have a demo project that we could play with? 🙏
> 
> In the future it would be really great to have iOS and Android parts of the `frameit` separated and have a `core` part being reused between those, wouldn't be? 😉

Thanks. Regarding the separation of iOS and Android, I was also considering this option and I was [asking about it in the original discussion](https://github.com/fastlane/fastlane/issues/14968). Eventually I decided for a single module. I think for example separating `deliver` and `supply` is logical, because App Store and Google Play are different, but in this case the task is practically the same, we just need to tell frameit what frames it should prefer, that's why I created the `use_platform` options and commands and making iOS default for the backward compatibility.

milhauscz committed:
Updated documentation - "Set of Android devices"

Co-Authored-By: Jan Piotrowski <piotrowski+github@gmail.com>

milhauscz committed:
Refactored checking frame colors in config_parser.rb according to @mollyIV's suggestion

milhauscz commented on frameit/lib/frameit/config_parser.rb:
```diff
@@ -103,12 +103,32 @@ def validate_key(key, value)
     end
+
+    def check_frame(value)
+      unless value.nil?
```
This is definitely more elegant and readable solution. I've implemented the changes [in the new commit](https://github.com/milhauscz/fastlane/commit/f6511282a916e259dc35e523dbdb26d88adb5c9a).

milhauscz committed:
Refactored REQUIRED_PRIORITY according to @mollyIV's suggestion

milhauscz committed:
Minor if refactorings according to @mollyIV's suggestion

milhauscz committed:
Removed a redundant function in editor

milhauscz committed:
Reverted a comment modification

milhauscz committed:
Refactoring of checking config / options

milhauscz commented on frameit/lib/frameit/config_parser.rb:
```diff
@@ -103,12 +103,32 @@ def validate_key(key, value)
+      when 'use_platform'
+        Platform.check_platform(value)
+      when 'force_device_type'
+        Device.check_device(value)
```
I agree `config_parser` is responsible for that, however the same option is also checked by `options`. And there is another difference, the `force_device_type` option supports both IDs and names, individual devices in the `Devices` module are objects and not strings. [I did some refactoring](https://github.com/milhauscz/fastlane/commit/d6b955eccfeb857070c0ad839caafb92d352065b), the `supported_device?` function uses `Device.find_device_by_id_or_name` and I created `Devices.all_device_names_without_apple` which can be used for printing devices to the error messages. All the functions have been moved to `config_parser` and `options` calls them. Also fixed some bugs from the previous commit and made all functions static.

milhauscz committed:
Fixes according to rubocop (added parenthesis to a method call and lowered perceived complexity of config_parser.validate_key)

milhauscz commented:
I've tried solving all the suggestions by adding new commits, please check it. Regarding the testing environment, [you can try this](https://drive.google.com/open?id=1oW3xyB6aOcTzAvWpW_mXXPcZaaHkyjn4). It contains three folders:

- **screengrab**: from a real Android project with Google Pixel 3 and Nexus 9 devices
- **screengrab_legacy_testing**: contains several iPhone screenshots where the `use_legacy` options can be tested, e. g. `fastlane run frameit "use_legacy_iphone_6s":true`, also contains a Samsung Galaxy S9 screenshot where the `force_device_type` gives it preference over the Pixel device; I just used some tablet screenshots and resized them without preserving the aspect ratio so that they match these devices, so ignore that resulting screenshots will look ugly
- **screengrab_old**: yet another test testing the `frame` option with a new color and combining Android and iOS platforms

milhauscz committed:
frames_generator will sort devices in JSONs alphabetically according to @janpio's suggestion

phikes commented:
If any help is needed here, let me know! Looking forward to this feature a lot :-) thanks for your work!

milhauscz committed:
Merge remote-tracking branch 'upstream/master'

# Conflicts:
#	frameit/lib/frameit/screenshot.rb

milhauscz committed:
Added new devices from PR #15373 as a fix for merge conflict in frameit/lib/frameit/screenshot.rb

milhauscz commented:
> If any help is needed here, let me know! Looking forward to this feature a lot :-) thanks for your work!

Thanks, currently everything is done on my side, frameit is working for both iOS and Android in a backward compatible way with the old frameit. I am waiting for the reviewers.

mollyIV commented:
Hey @milhauscz 👋 

Thanks a bunch for making the code review changes, the code looks 💯 I would like to move to manual testing part now. 

> Regarding the testing environment, you can try this.

I requested the access to this project, could you please grant it? Thanks!

milhauscz commented:
> Hey @milhauscz 👋
> 
> Thanks a bunch for making the code review changes, the code looks 💯 I would like to move to manual testing part now.
> 
> > Regarding the testing environment, you can try this.
> 
> I requested the access to this project, could you please grant it? Thanks!

Hi, I have granted the request and here is also a new link which should be publicly accessible: https://drive.google.com/file/d/1oW3xyB6aOcTzAvWpW_mXXPcZaaHkyjn4/view?usp=sharing

milhauscz committed:
Updated the frames generator to correctly rename iPads in the latest Facebook devices
Updated device_spec to count with newer iPhone 11 Pro
Added new devices: Google Pixel 4, Google Pixel 4 XL, Samsung Galaxy Note 10, Samsung Galaxy Note 10+, Samsung Galaxy S10, Samsung Galaxy S10+, Xiaomi Mi Mix Alpha, iPhone 11, iPhone 11 Pro, iPhone 11 Pro Max, iPad 10.2, iPad Air 2019, iPad Mini 2019

milhauscz committed:
Added missing number for iPhone 11

milhauscz committed:
Increased Pixel 4's priority to be higher than Galaxy S10
Fixed name for iPad Pro 12.9 3rd gen.
Fixed device name sanitization in offsets.rb
Created legacy options for iPhone XR, XS and XS Max

milhauscz commented:
I've added some new devices based on @mollyIV's request (Google Pixel 4, Google Pixel 4 XL, Samsung Galaxy Note 10, Samsung Galaxy Note 10+, Samsung Galaxy S10, Samsung Galaxy S10+, Xiaomi Mi Mix Alpha, iPhone 11, iPhone 11 Pro, iPhone 11 Pro Max, iPad 10.2, iPad Air 2019, iPad Mini 2019).
Did some manual and automatic tests and fixed bugs. Should be ready now. The automatic tests here in GitHub are failing, but I think the failure is not caused by my changes, I am able to successfully run `gem build fastlane.gemspec` in my machine.

mollyIV commented:
I was finally able to test the provided sample project. 
Overall it works really great 💯 

I got a few warnings though:

```bash
$ bundle exec fastlane frameit
```

```bash
[⠋] Framing screenshot './en-US/iphone_6splus_7plus_offline_view_1558018072937.png' [09:18:36]: Unfortunately device type 'Apple iPhone 7 Plus' is not available in Sunrise Gold, falling back to Matte Black...
[⠋] Framing screenshot './en-US/iphone_5s_se_annotation_1558018050803.png' [09:18:42]: Unfortunately device type 'Apple iPhone SE' is not available in Sunrise Gold, falling back to Space Gray...
```

```bash
[⠋] Framing screenshot './en-US/images/screenshots/android_tablet_1_Dashboard_1564152647944.png' [09:23:29]: Unfortunately device type 'Nexus 9' is not available in Clearly White, falling back to default...
```

```bash
[⠋] Framing screenshot './en-US/tablet_getting_started_1558018068730.png' [09:30:24]: Unfortunately device type 'Nexus 9' is not available in Sunrise Gold, falling back to default...
```

It looks like these ☝️relate to `Framefile.json` config file, so all good 👍 

I wanted to say thank you @milhauscz for you out-standing work on this feature and apologise that it took me soo long to review and test it out  🙇 

I think this PR is still **blocked by** https://github.com/fastlane/frameit-frames/pull/10 PR (introducing the new frames) though. 

I'd also like to wait for a review and testing from another member of `fastlane` team (just to double check everything's fine - this set of changes is quite huge) 😊 

I am very excited about supporting Android screenshots 📸 

Thank you!

milhauscz commented:
> I was finally able to test the provided sample project.
> Overall it works really great 💯
> 
> I got a few warnings though:
> 
> ```shell
> $ bundle exec fastlane frameit
> ```
> 
> ```shell
> [⠋] Framing screenshot './en-US/iphone_6splus_7plus_offline_view_1558018072937.png' [09:18:36]: Unfortunately device type 'Apple iPhone 7 Plus' is not available in Sunrise Gold, falling back to Matte Black...
> [⠋] Framing screenshot './en-US/iphone_5s_se_annotation_1558018050803.png' [09:18:42]: Unfortunately device type 'Apple iPhone SE' is not available in Sunrise Gold, falling back to Space Gray...
> ```
> 
> ```shell
> [⠋] Framing screenshot './en-US/images/screenshots/android_tablet_1_Dashboard_1564152647944.png' [09:23:29]: Unfortunately device type 'Nexus 9' is not available in Clearly White, falling back to default...
> ```
> 
> ```shell
> [⠋] Framing screenshot './en-US/tablet_getting_started_1558018068730.png' [09:30:24]: Unfortunately device type 'Nexus 9' is not available in Sunrise Gold, falling back to default...
> ```
> 
> It looks like these ☝️relate to `Framefile.json` config file, so all good 👍
> 
> I wanted to say thank you @milhauscz for you out-standing work on this feature and apologise that it took me soo long to review and test it out 🙇
> 
> I think this PR is still **blocked by** [fastlane/frameit-frames#10](https://github.com/fastlane/frameit-frames/pull/10) PR (introducing the new frames) though.
> 
> I'd also like to wait for a review and testing from another member of `fastlane` team (just to double check everything's fine - this set of changes is quite huge) 😊
> 
> I am very excited about supporting Android screenshots 📸
> 
> Thank you!

Thanks for checking this PR. Yes, you're right, the warning messages are caused by the configuration file. This actually tests whether the new fallback to default color works. Looking forward for a review from another member.

Also I would like to ask if someone could help me checking the cause of the failed automatic check (ci/circleci: Execute modules load up tests). As I mentioned I think this was not caused by my changes.

joshdholtz commented:
Heyyy 👋 I've been sidetracked with a lot of other Android related things for `supply` and `screengrab` but looking forward to getting to test this one out soon 😊

shawnlius commented:
Hey @milhauscz,  thanks for the nice work! Can't wait for the merge!

Just want to point out, I do get an error when calling frameit() from Fastfile:

```
Action 'frameit' isn't known to support operating system 'android'.
```

I think adding `:android` here will do:

https://github.com/fastlane/fastlane/blob/1bb7280cf86e18abc93841dd941aa958255694f0/fastlane/lib/fastlane/actions/frame_screenshots.rb#L60

janpio committed:
Merge branch 'master' into master

janpio commented:
I did the rebase in the branch.

@shawnlius has the right idea regarding the platform.

kdo1234 commented:
It will be great when this is merged!

patreu22 commented:
Does anybody know when this PR is scheduled for merging? Would be great to have it in the next Fastlane release and should be ready to ship, right?

janpio commented:
Obviously not ready to merge if there is feedback about things that are broken: https://github.com/fastlane/fastlane/pull/15087#issuecomment-554280006

milhauscz committed:
Merge remote-tracking branch 'upstream/master'

milhauscz committed:
Added Android as a supported platform in frame_screenshots

milhauscz committed:
Merge remote-tracking branch 'origin/master'

milhauscz commented:
> Hey @milhauscz, thanks for the nice work! Can't wait for the merge!
> 
> Just want to point out, I do get an error when calling frameit() from Fastfile:
> 
> ```
> Action 'frameit' isn't known to support operating system 'android'.
> ```
> 
> I think adding `:android` here will do:
> 
> https://github.com/fastlane/fastlane/blob/1bb7280cf86e18abc93841dd941aa958255694f0/fastlane/lib/fastlane/actions/frame_screenshots.rb#L60

OK, fixed and merged upstream/master without conflicts. Everything is done on my side so hopefully this PR can be merged soon.

milhauscz committed:
Failed test fix - do not use frameit for checking unsupported platform since it already supports Android; replaced by team_name

milhauscz committed:
Failed test fix - do not use frameit for checking unsupported platform since it already supports Android; replaced by team_name

milhauscz commented:
I've also fixed a failing test, which was testing the error message which should be shown if an action is run inside an environment with an unsupported platform and which unfortunately used "frameit" with "android". Replaced by the "team_name" action which doesn't support Android.

So now all is ready for a merge.

milhauscz commented:
@mollyIV @janpio @joshdholtz Kindly asking for a review, thanks.

joshdholtz commented:
Aye aye! This one is on my list to look at 😊 I’m pretty excited for it just gotta squash a few other tasks I’m working on first 💪

mollyIV reviewed (approved):
Hey @milhauscz 👋 

Thanks again for all your hard work and being patient 🙇 I took the newest changes for a spin and all frames look good 🙌 The only thing is that I got a warning when framing the screenshots from `screengrab` catalog from provided sample project:

```bash
[⠇] Framing screenshot './en-US/images/screenshots/android_phone_2_Get_Started_1564149855906.png' convert: invalid profile length `/var/folders/n0/8g1hyb49749b08r3c1yf45pw0000gq/T/mini_magick20200227-3595-12qbi5t.png' @ warning/png.c/MagickPNGWarningHandler/1747.
```

⚠️  Let's not forget [🔑Add Android device frames](https://github.com/fastlane/frameit-frames/pull/10) pull request should be merged before this one though.

Thanks again, let's 🚢 🚢

joshdholtz reviewed (approved):
This is sooooooo 🔥 Merging this now and shipping this today! Thanks so much for the contribution ❤️
